framework/src/variables/MooseVariableDataFV.C

```diff
@@ -1143,6 +1143,55 @@ MooseVariableDataFV<OutputType>::dofValuesDuDotDotDu() const
   return _dof_du_dotdot_du;
 }
 
+template <typename OutputType>
+const typename MooseVariableDataFV<OutputType>::FieldVariableValue &
+MooseVariableDataFV<OutputType>::vectorTagValue(TagID tag) const
+{
+  // we need to check special tags until we use tagging for all evaluations
+  // otherwise we will have redundant evaluations.
+  if (tag == _subproblem.getVectorTagID(Moose::SOLUTION_TAG))
+    return sln(Moose::Current);
+  else if (_subproblem.vectorTagExists(Moose::OLD_SOLUTION_TAG) &&
+           tag == _subproblem.getVectorTagID(Moose::OLD_SOLUTION_TAG))
+    return sln(Moose::Old);
+  else if (_subproblem.vectorTagExists(Moose::OLDER_SOLUTION_TAG) &&
+           tag == _subproblem.getVectorTagID(Moose::OLDER_SOLUTION_TAG))
+    return sln(Moose::Older);
```

Comment:
It would be helpful if you could explain why we need to special handles here for `OLDER_SOLUTION_TAG` and others. I feel like we want to handle them in an uniform way if that is possible

Reply:
I am just mimicking what we already have in `MooseVariableData`. But there is a problem with the tagging system linked to this code. We size `_vector_tag_u` and related data members in the variable constructors based off the current number of vector tags in the `_subproblem` at that time. However, older solution states may not be requested until later. This can cause indexing out-of-bounds in `_vector_tag_u` unless you use things like `sln(Moose::OLD)` and `sln(Moose::OLDER)`. We could change 
```c++
  else
  {
    _need_vector_tag_u[tag] = true;

    if (_sys.hasVector(tag) && tag < _vector_tag_u.size())
      return _vector_tag_u[tag];
    else
      mooseError("Tag is not associated with any vector or there is not any data for tag ",
                 tag,
                 " for variable ",
                 _var.name());
  }
```
to something like
```c++
  else
  {
    if (tag >= _vector_tag_u.size())
      {
        _need_vector_tag_u.resize(tag + 1, false);
        _vector_tag_u.resize(tag + 1);
      }
      
    _need_vector_tag_u[tag] = true;

    if (_sys.hasVector(tag))
      return _vector_tag_u[tag];
    else
      mooseError("Tag ", tag, " is not associated with any vector for variable ",
                 _var.name());
  }
```
and then we could just `_vector_tag_u` for everything and not do special dispatch for certain solution vector tags. What do you think?

Reply:
I like this idea better.

Reply:
This will make code clearer and simpler

Reply:
this has not been simpler ... this has turned into a monstrous task :frowning_face: